go/pserver/etcd_lister.go

```diff
@@ -0,0 +1,120 @@
+package pserver
+
+import (
+	"time"
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	DefaultEtcdTimeout time.Duration = time.Second * time.Duration(5)
+)
+
+type pserverEtcdLister struct {
+	client    *clientv3.Client
+	timeout   time.Duration
+	endpoints []string
+}
+
+// read ps desired number from etcd.
+func(p *pserverEtcdLister) desired() int {
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+		resp, err := p.client.Get(ctx, DefaultPsDesiredPath)
+		cancel()
+		if err != nil {
+			log.Errorf("Get ps dresire number failed! recnnectiong..., %v", err)
+			time.Sleep(p.timeout)
+			continue
+		}
+
+		kvs := resp.Kvs
+		if len(kvs) == 0 {
+			log.Infoln("Waiting for ps desired registered ...")
+			time.Sleep(p.timeout)
+			continue
+		}
+
+		psDesired, err := strconv.Atoi(string(resp.Kvs[0].Value))
+		if err != nil {
+			log.Errorf("psDesired %s invalid %v", psDesired, err)
+			time.Sleep(p.timeout)
+			continue
+		}
+
+		log.Debugf("Get psDesired number: %d\n", psDesired)
+		return psDesired
+	}
+}
+
+func(p *pserverEtcdLister) List() []Server {
+	psDesired := p.desired()
+
+	servers := make([]Server, psDesired)
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+		for i := 0; i < psDesired; i++ {
```

Comment:
This can be batched into a single transaction, here is an related example: https://github.com/PaddlePaddle/Paddle/blob/develop/go/master/etcd_client.go#L121
Something like:
```Go
resp, err := e.client.Txn(ctx).If().Then(get0, get1, get2, get3).Commit()
```
Feel free to improve it later, not a merge blocker.